File: p2p-ws-sdk-dart/lib/src/frame.dart
import "dart:typed_data";

class WireHeader {
  final int length;
  final int magic;
  final int version;
  final int flags;

  const WireHeader(this.length, this.magic, this.version, this.flags);
}

class DecodedFrame {
  final WireHeader header;
  final Uint8List cipherPayload;

  const DecodedFrame(this.header, this.cipherPayload);
}

Uint8List encodeFrame(WireHeader h, Uint8List payload) {
  final out = Uint8List(8 + payload.length);
  final bd = ByteData.sublistView(out);
  bd.setUint32(0, h.length, Endian.big);
  bd.setUint16(4, h.magic & 0xFFFF, Endian.big);
  bd.setUint8(6, h.version & 0xFF);
  bd.setUint8(7, h.flags & 0xFF);
  out.setRange(8, out.length, payload);
  return out;
}

DecodedFrame decodeFrame(Uint8List frame) {
  if (frame.length < 8) {
    throw FormatException("frame too short");
  }
  final bd = ByteData.sublistView(frame);
  final length = bd.getUint32(0, Endian.big);
  final magic = bd.getUint16(4, Endian.big);
  final version = bd.getUint8(6);
  final flags = bd.getUint8(7);
  final payload = Uint8List.sublistView(frame, 8);
  if ((flags & 0x04) != 0 && length != payload.length) {
    throw FormatException("length check failed");
  }
  return DecodedFrame(WireHeader(length, magic, version, flags), payload);
}

File: p2p-ws-sdk-ts/src/frame.ts
export type WireHeader = {
  length: number
  magic: number
  version: number
  flags: number
}

export type WireFrame = {
  header: WireHeader
  cipherPayload: Uint8Array
}

export const HEADER_LEN = 8

export function decodeFrame(wsBinaryPayload: Uint8Array): WireFrame {
  if (wsBinaryPayload.length < HEADER_LEN) {
    throw new Error("invalid frame")
  }
  const dv = new DataView(wsBinaryPayload.buffer, wsBinaryPayload.byteOffset, wsBinaryPayload.byteLength)
  const length = dv.getUint32(0, false)
  const magic = dv.getUint16(4, false)
  const version = dv.getUint8(6)
  const flags = dv.getUint8(7)
  const cipherPayload = wsBinaryPayload.subarray(HEADER_LEN)
  return { header: { length, magic, version, flags }, cipherPayload }
}

export function encodeFrame(header: WireHeader, cipherPayload: Uint8Array): Uint8Array {
  const out = new Uint8Array(HEADER_LEN + cipherPayload.length)
  const dv = new DataView(out.buffer, out.byteOffset, out.byteLength)
  dv.setUint32(0, header.length >>> 0, false)
  dv.setUint16(4, header.magic & 0xffff, false)
  dv.setUint8(6, header.version & 0xff)
  dv.setUint8(7, header.flags & 0xff)
  out.set(cipherPayload, HEADER_LEN)
  return out
}

